subsys/usb/class/netusb/function_ecm.c

```diff
@@ -141,6 +149,16 @@ static void ecm_bulk_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 		in_pkt = pkt;
 	}
 
+	/* The ECM spec says that the end of a packet is marked by
+	 * a short frame.  If the packet is a multiple of the endpoint
+	 * size then the host should send a zero length packet.
+	 * Linux, however, sends a one byte packet instead.  Handle by
+	 * checking the IP header length and dropping the extra byte.
+	 */
+	if (len == 1 && ecm_have_payload(in_pkt)) {
+		len = 0;
+	}
+
 	if (!net_pkt_append_all(in_pkt, len, buffer, K_FOREVER)) {
 		SYS_LOG_ERR("Error appending data to pkt: %p", in_pkt);
```

Comment:
I see that net_pkt_append() is quite expensive, @jukkar does it make sense to have following patch:
```

diff --git a/subsys/net/ip/net_pkt.c b/subsys/net/ip/net_pkt.c
index 6a062174d..6599c93c6 100644
--- a/subsys/net/ip/net_pkt.c
+++ b/subsys/net/ip/net_pkt.c
@@ -1203,7 +1203,7 @@ u16_t net_pkt_append(struct net_pkt *pkt, u16_t len, const u8_t *data,
        struct net_context *ctx = NULL;
        u16_t max_len, appended;
 
-       if (!pkt || !data) {
+       if (!pkt || !data || !len) {
                return 0;
    }
```

Reply:
> I see that net_pkt_append() is quite expensive, @jukkar does it make sense to have following patch

Looks quite sane thing to do. Could you send a PR for that?

Reply:
Sent PR #6394